src/api/Controllers/CoursesController.cs

```diff
@@ -62,9 +63,12 @@ public IActionResult Index([FromBody]IList<Course> courses)
                     MakeProvidersDistinctReferences(ref courses);
                     MakeRoutesDistinctReferences(ref courses);
 
+                    // these methods may add new locations/subjects. We need to save these changes before adding courses
+                    // otherwise we will end up with a PK constraint exception
                     AssociateWithLocations(ref courses);
                     AssociateWithSubjects(ref courses);
-
+                    _context.SaveChanges();
```

Comment:
I dont think there is a need to call it twice, as the final call should sort it out as it would of had the same reference